lib/arm/crc32_impl.h

```diff
@@ -151,12 +152,66 @@ crc32_pmull_aligned(u32 remainder, const uint8x16_t *p, size_t nr_segs)
 #include "../crc32_vec_template.h"
 #endif /* PMULL implementation */
 
+#undef DISPATCH_INSTR
+#if (defined(__ARM_FEATURE_CRC32) ||	\
+    (ARM_CPU_FEATURES_ENABLED &&	\
+     COMPILER_SUPPORTS_CRC32_TARGET_INTRINSICS)) && \
+     /* not yet tested on big endian, probably needs changes to work there */ \
+    (defined(__BYTE_ORDER__) && __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__)
+#  define FUNCNAME		crc32_instr
+#  define FUNCNAME_ALIGNED	crc32_instr_aligned
+#  define FUNCNAME_SLICE1	crc32_instr_slice1
+#  ifdef __ARM_FEATURE_CRC32
+#    define ATTRIBUTES
+#    define DEFAULT_IMPL	crc32_instr
+#  else
+#    define __ARM_FEATURE_CRC32	1
```

Comment:
What is this line for?

Reply:
The header only defines `__crc32..` (as equivalent to `__builtin_...crc32..`) when that is defined (i.e. when the target feature is enabled, which we're doing here via an attribute, but that doesn't set these defines as that happens after the preprocessor).

As I understand, `__crc32..` is the public API.

Reply:
It works for me even I delete that line (I'm using gcc 8.2.1 on Linux on AArch64).

Normally the CPU feature macros like `__ARM_FEATURE_*`, `__AVX2__`, etc. are meant to be defined by the compiler based on the compilation flags, not by the user.  And intrinsics are available in `target` attribute functions regardless of these, though sometimes not with old compilers.  The way I've handled this in libdeflate elsewhere is just to disable the code if the compiler is too old.  I don't want to go too far into the weeds of trying to make all the optimizations work with old compilers.

I don't see why we shouldn't do the same here.  Can you explain what compiler versions need this, exactly?

Reply:
> intrinsics are available in target attribute functions regardless of these

That's not the case with clang.

https://github.com/llvm/llvm-project/blob/18b0c40bc5dbaf25157a838701b2aba8ba8440db/clang/lib/Headers/arm_acle.h#L558

(this is the newest revision)

Reply:
Well, I guess we can do it then, but please leave a brief comment explaining that it's a workaround for clang.  Can you also check whether there's any clang bug report for this, and if not, file one?  A more minimal example:

```
#include <arm_neon.h>
#include <arm_acle.h>

poly128_t __attribute__((target("crypto")))
crypto_insn(poly64_t a)
{
        return vmull_p64(a, a);
}

uint32_t __attribute__((target("crc")))
crc32_insn(uint32_t crc, uint8_t byte)
{
        return __crc32b(crc, byte);
}
```

With default flags, i.e. no crypto or crc enabled, clang can compile the first function, but not the second function.  gcc can compile both.  (Well, for gcc you have to prepend "+" to "crypto" and "crc".)